### ISSUE TITLE
float (operator) float works sometimes...

### ISSUE THREAD

**Issue by 3FLLC** [nielsAD/lape#76]
the following code, run, works, run, error. found if I do a writeln(value) it works every time if that helps?

  Assert(1 + 1 = 2,        'Int Add Int');
  Assert(1.1 + 2 = 3.1,    'Float Add Int');
  Assert(1 + 2.1 = 3.1,    'Int Add Float');
// fails some times:
  Assert(1.1 + 2.1 = 3.2,  'Float Add Float');
  Assert('1' + '2' = '12', 'String Add String');

  Assert(1 - 1 = 0,        'Int Subtract Int');
  Assert(2.0 - 1.0 = 1.0,  'Float Subtract Float');
// fails some times:
  Assert(2.1 - 1.1 = 1.0,  'Float Subtract Float');

  Assert(2 * 3 = 6,        'Int Multiply Int');
// fails some times:
  Assert(2.1 * 3.1 = 6.51, 'Float Multiply Float');
  Assert(2 * 3.1 = 6.2,    'Int Multiply Float');
  Assert(3.1 * 2 = 6.2,    'Float Multiply Int');

  Assert(6 / 2 = 3,        'Int Divide Int = Int');
// fails some times:
  Assert(6.51 / 3.1 = 2.1, 'Float Divide Float');
  Assert(1 / 2 = 0.5,      'Int Divide Int = Float');
  Assert(1 / 2.5 = 0.4,    'Int Divide Float');

again, adding a writeln(of the formula before it) makes it work every time. Integers works perfectly, only floats +, -, *, / floats fails.

**Comment by WarPie** [comment]
Comparing floats does not guarantee equality due to precision errors, this seems to follow FPC
http://ideone.com/lXaCeH

**Comment by 3FLLC** [comment]
However, if you do the writeln(float operator float = expected result) it returns true for the next line doing the actual assert... do writeln or tostring or something is cleaning up an issue... that is the bug.

**Comment by nielsAD** [comment]
Can you give an example of where *WriteLn* would show *True* but *Assert* fails? As @WarPie explained, most of these comparisons are expected to fail because of the [representation of floating point numbers](http://rvelthuis.de/articles/articles-floats.html). *FloatToStr* only prints up to 15 decimals, which might not be the entire number.

```
{$assertions on}

begin
  WriteLn(1.1 + 2.1, ' = 3.2 -> ', 1.1 + 2.1 = 3.2);
  WriteLn(Format('%.18f', [1.1 + 2.1]));
  Assert(1.1 + 2.1 = 3.2);
end;
```

**Comment by 3FLLC** [comment]
Here is the source:
  Assert(1 + 1 = 2,        'Int Add Int');
  Assert(1.1 + 2 = 3.1,    'Float Add Int');
  Assert(1 + 2.1 = 3.1,    'Int Add Float');
  Assert(1.1 + 2.1 = 3.2,  'Float Add Float');
  Assert('1' + '2' = '12', 'String Add String');

run:
>>> Runtime error: "Assertion failure: Float Add Float" at line 4, column 3

change to:
  Assert(1 + 1 = 2,        'Int Add Int');
  Assert(1.1 + 2 = 3.1,    'Float Add Int');
  Assert(1 + 2.1 = 3.1,    'Int Add Float');
Writeln(1.1+2.1);
  Assert(1.1 + 2.1 = 3.2,  'Float Add Float');
  Assert('1' + '2' = '12', 'String Add String');


run:
[root@db mp2]# mp p
3.2
[root@db mp2]# 

simply putting that written, for someone reason makes the float op float comparison work. (this is the easiest one i have to document, I have some extremely weird issues that by calling writeln(something) makes memory leaks go away, bogus exceptions go away, etc. This is why i am making the most robust self-test I can to figure out wtf.

Thanks,
Ozz

**Comment by 3FLLC** [comment]
PS I do understand the FPC example - I am not denying how floats work, but, why adding writeln(formula) makes it work. ;-)

**Comment by nielsAD** [comment]
I can't reproduce this. Are you testing with the latest version?

**Comment by 3FLLC** [comment]
My version is very close to the latest (within reason). I have many changes (as my LaPe is not "Pascal" ISO compliant)... however, as far as I can see, this is not an area I changed. (change not changed "internal" methods).

**Comment by WarPie** [comment]
I can reproduce in Simba using an older lape, I can however not reproduce in the little lape "editor". It might be optimization related.

**Comment by 3FLLC** [comment]
Thanks for your time. I guess ISSUE is the wrong category, more of us, huh? I found it as I am writing a full end to end dunit type solution - and the sample for Delphi had those equations, LaPe/FPC does as you guys expect... problem was, adding the writeln(formula) to see wtf, fixed it... thus my inquiry.

PS. My original LAPE source was from google I think, then my next upgrade was Simba - then I finally found Niels on github - but I have 21 months of changes to my core.

Ozz
